src/wp-includes/blocks.php

```diff
@@ -760,6 +760,29 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
+function insert_hooked_blocks( $block, $parent = null, $block_index = null, $chunk_index = null ) {
+	$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+		$hooked_block = array(
+			'blockName'    => $hooked_block_type,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+		if ( 'before' === $relative_position ) {
+			insert_inner_block( $parent, $block_index, $chunk_index, $hooked_block );
```

Comment:
I had a realization last night that `before` (and probably `after`) insertion isn't going to work like this in combination with tree traversal and serialization 😕 The reason is that we're iterating over `parent`'s `innerBlocks` array in `traverse_and_serialize_blocks` when we encounter this line, which then attempts to modify that array (by inserting another element before or after the current one, respectively).

I came up with an idea how to fix that. The good news is that if it works, we'll be able to get rid of our `$chunk_index` arithmetics, and even remove `insert_inner_block` and friends. The bad news is that I'll have to change the function signature of the callback in `traverse_and_serialize_blocks` once again 😬